src/gen/sanitize_int.rs

```diff
@@ -0,0 +1,274 @@
+use super::rangeset::RangeSet;
+use super::validation_error::{atom, ErrorCode, ValidationErr};
+use crate::allocator::{Allocator, AtomBuf, NodePtr, SExp};
+use crate::run_program::STRICT_MODE;
+
+fn count_zeros_impl(buf: &[u8]) -> usize {
+    let mut ret: usize = 0;
+    while ret < buf.len() && buf[ret] == 0 {
+        ret += 1;
+    }
+    ret as usize
+}
+
+// this *just* counts zeros. Note that not all zeros are redundant. the last
+// zero byte may be necessary. This has to be fixed up by the caller
+fn count_zeros(a: &Allocator, range_cache: &mut RangeSet, atom_range: &AtomBuf) -> usize {
+    if atom_range.is_empty() {
+        return 0;
+    }
+
+    // early exit for common case
+    if atom_range.len() < 20 {
+        return count_zeros_impl(a.buf(atom_range));
+    }
+
+    let range = atom_range.idx_range();
+    let ranges = range_cache.not_overlapping(range.0, range.1);
+
+    // an empty range means the entire range we passed in is all zeros
+    if ranges.is_empty() {
+        return atom_range.len();
+    }
+
+    let full_buffer = a.buf(atom_range);
+
+    let mut start = range.0;
+    let mut ret = 0_usize;
+    for r in ranges {
+        debug_assert!(r.0 >= range.0);
+        debug_assert!(r.1 <= range.1);
+        debug_assert!(r.0 < r.1);
+
+        if r.0 != start {
+            // skip this range of known zero-bytes
+            ret += (r.0 - start) as usize;
+        }
+        let buf = &full_buffer[(r.0 - range.0) as usize..(r.1 - range.0) as usize];
+        let zeros = count_zeros_impl(buf);
+        ret += zeros;
+        if zeros < buf.len() {
+            break;
+        }
+        start = r.1;
+    }
+
+    debug_assert!(start <= range.1);
+
+    if ret > 0 {
+        range_cache.add(range.0, range.0 + ret as u32);
+    }
+    ret
+}
+
+pub fn sanitize_integer<'a>(
+    a: &'a Allocator,
+    n: NodePtr,
+    max_size: usize,
+    code: ErrorCode,
+    range_cache: &mut RangeSet,
+    flags: u32,
+) -> Result<&'a [u8], ValidationErr> {
+    assert!(max_size <= 8);
+
+    let buf = atom(a, n, code)?;
+
+    if buf.is_empty() {
+        return Ok(&[]);
+    }
+
+    // if the top bit is set, it's a negative number
+    if (buf[0] & 0x80) != 0 {
+        return Err(ValidationErr(n, ErrorCode::NegativeAmount));
+    }
+
+    let atom_range = match a.sexp(n) {
+        SExp::Atom(b) => b,
+        _ => {
+            panic!("unreachable");
+        }
+    };
+
+    // strip redundant leading zeros
+    let mut i = count_zeros(a, range_cache, &atom_range);
```

Comment:
Any reason this doesn't update the range cache when it counts them? It has it borrowed `&mut`, right? Does it just make the code too messy?

Reply:
Oh I see, it doesn't have `atom_range` available.